clippy_lints/src/feature_name.rs

```diff
@@ -0,0 +1,172 @@
+                    let i = PREFIXES.partition_point(|prefix| prefix < &feature.as_str());
+                    if i > 0 && feature.starts_with(PREFIXES[i - 1]) {
+                        Some(PREFIXES[i - 1])
+                    } else {
+                        None
+                    }
+                };
+                if let Some(prefix) = prefix_opt {
+                    if is_negative_prefix(prefix) {
+                        span_lint_and_help(
+                            cx,
+                            NEGATIVE_FEATURE_NAMES,
+                            DUMMY_SP,
+                            &format!(
+                                "the \"{}\" prefix in the feature name \"{}\" is negative",
+                                prefix, feature
+                            ),
+                            None,
+                            &format!(
+                                "consider renaming the feature to \"{}\", but make sure the feature adds functionality",
+                                feature.strip_prefix(prefix).unwrap()
+                            ),
+                        );
+                    } else {
+                        span_lint_and_help(
+                            cx,
+                            REDUNDANT_FEATURE_NAMES,
+                            DUMMY_SP,
+                            &format!(
+                                "the \"{}\" prefix in the feature name \"{}\" is redundant",
+                                prefix, feature
+                            ),
+                            None,
+                            &format!(
+                                "consider renaming the feature to \"{}\"",
+                                feature.strip_prefix(prefix).unwrap()
+                            ),
+                        );
+                    }
+                }
+
+                let suffix_opt: Option<&str> = {
+                    let i = SUFFIXES.partition_point(|suffix| {
+                        suffix.bytes().rev().cmp(feature.bytes().rev()) == std::cmp::Ordering::Less
+                    });
+                    if i > 0 && feature.ends_with(SUFFIXES[i - 1]) {
+                        Some(SUFFIXES[i - 1])
+                    } else {
+                        None
+                    }
+                };
+                if let Some(suffix) = suffix_opt {
+                    span_lint_and_help(
+                        cx,
+                        REDUNDANT_FEATURE_NAMES,
+                        DUMMY_SP,
+                        &format!(
+                            "the \"{}\" suffix in the feature name \"{}\" is redundant",
+                            suffix, feature
+                        ),
+                        None,
+                        &format!(
+                            "consider renaming the feature to \"{}\"",
+                            feature.strip_suffix(suffix).unwrap()
+                        ),
+                    );
```

Comment:
I still would like to see all the `span_lint_and_help` cases collapsed into one function to reduce redundancy.
```rust
fn lint(cx, feature: &str, prefix: &str, is_suffix: bool, negative: bool) { .. }
```